lib/extensions.js

```diff
@@ -106,38 +113,112 @@ module.exports = function(realmConstructor, context) {
 
     //Add static methods to the Realm object
     Object.defineProperties(realmConstructor, getOwnPropertyDescriptors({
-
-        open(config) {
+        open: function (config) {
             // If no config is defined, we should just open the default realm
-            if (config === undefined) { config = {}; }
+            if (config === undefined) {
+                config = {};
+            }
 
             // For local Realms we open the Realm and return it in a resolved Promise.
             if (!("sync" in config)) {
-                let promise = Promise.resolve(new realmConstructor(config));
-                promise.progress = (callback) => { };
-                return promise;
+                return openLocalRealm(realmConstructor, config);
             }
 
-            let syncSession;
-            let promise = new Promise((resolve, reject) => {
-                syncSession = realmConstructor._asyncOpen(config, (realm, error) => {
-                    if (error) {
-                        setTimeout(() => { reject(error); }, 1);
-                    }
-                    else {
-                        setTimeout(() => { resolve(realm); }, 1);
-                    }
-                });
-            });
+            // Determine if we are opening an existing Realm or not.
+            let behavior = realmConstructor.exists(config) ? "existingRealmFileBehavior" : "newRealmFileBehavior";
+            
+            // Define how the Realm file is opened
+            let openLocalRealmImmediately = false; // Default is syncBeforeOpen
+            if (config.sync[behavior] !== undefined) {
+                const type = config.sync[behavior].type;
+                switch (type) {
+                    case 'syncBeforeOpen': 
+                        openLocalRealmImmediately = false; 
+                        break;
+                    case 'openImmediately': 
+                        openLocalRealmImmediately = true;
+                        break;
+                    default:
+                        throw Error("Invalid type: '" + type + "'. Only 'syncBeforeOpen' and 'openImmediately' is allowed.")
+                }
+            }
+
+            // Configure timeout
+            let timeOut = 0x7FFFFFFF; // Max value allowed by setTimeOut()
```

Comment:
This doesn't appear to really simplify the code as compared to just skipping the whole timeout thing if the timeout is undefined.